GGRC-33 Add logic to recreate lost revisions

alieh-rymasheuski opened the pull request:
This PR adds logic to insert lost revisions:
1. For a lost object with no "deleted" revision, create a copy of this object's last revision with `action="deleted"`.
2. For an object with no revisions and `created_at == updated_at`, create a new revision with `action="created"` and `content=obj.log_json()`.
3. For an object with no revisions and `created_at != updated_at`, create a new revision with `action="modified"` and `content=obj.log_json()`.

alieh-rymasheuski commented:
`revisions.created_at` and `revisions.updated_at` get invalid value `0000-00-00 ...` for the created `action="deleted"` revisions. It's connected with date manipulation bugs in MySQL v5.5.

Smotko commented:
@alieh-rymasheuski I'm guessing it's this is the query that doesn't insert the correct statements: https://github.com/google/ggrc-core/pull/4802/files#diff-daf6da2fa13feea3fb2bda51c7cf1d12R121 It's probably because you are removing the two columns from the insert select statement:

```
INSERT INTO revisions (event_id, action, content, context_id, destination_id, destination_type, modified_by_id, resource_id, resource_type, source_id, source_type)
```

Have you tried adding the columns to the INSERT statement and setting some default values for it in the SELECT?

Smotko commented:
I have received the following error when running this on the db from grc-test:

```
IntegrityError: (IntegrityError) (1452, 'Cannot add or update a child row: a foreign key constraint fails
(`ggrcdev`.`revisions`, CONSTRAINT `fk_revisions_contexts` FOREIGN KEY (`context_id`) REFERENCES `contexts` (`id`))') 'INSERT INTO revisions (resource_id, resource_type, event_id, action, content, source_type, source_id, destination_type, destination_id, created_at, modified_by_id, updated_at, context_id) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, current_timestamp, %s, current_timestamp, %s)' ((145L, 'UserRole', 53283L, 'created', '{"display_name": "aa@aa.com <-> ProgramReader", "context_id": 527, "created_at": "2015-07-17T20:44:45", "updated_at": "2015-07-17T20:44:45", "role_id": 3, "modified_by_id": null, "person_id": 10, "modified_by": null, "id": 145}', None, None, None, None, 150L, 527L), (146L, 'UserRole', 53283L, 'created', '{"display_name": "aa@aa.com <-> ProgramEditor", "context_id": 527, "created_at": "2015-07-17T20:44:45", "updated_at": "2015-07-17T20:44:45", "role_id": 2, "modified_by_id": null, "person_id": 10, "modified_by": null, "id": 146}', None, None, None, None, 150L, 527L), (147L, 'UserRole', 53283L, 'created', '{"display_name": "user1@ggrc.com <-> ProgramOwner", "context_id": 527, "created_at": "2015-07-17T20:44:45", "updated_at": "2015-07-17T20:44:45", "role_id": 1, "modified_by_id": null, "person_id": 31, "modified_by": null, "id": 147}', None, None, None, None, 150L, 527L), (152L, 'UserRole', 53283L, 'created', '{"display_name": "aa@aa.com <-> ProgramReader", "context_id": 537, "created_at": "2015-07-17T20:45:54", "updated_at": "2015-07-17T20:45:54", "role_id": 3, "modified_by_id": null, "person_id": 10, "modified_by": null, "id": 152}', None, None, None, None, 150L, 537L), (153L, 'UserRole', 53283L, 'created', '{"display_name": "aa@aa.com <-> ProgramEditor", "context_id": 537, "created_at": "2015-07-17T20:45:54", "updated_at": "2015-07-17T20:45:54", "role_id": 2, "modified_by_id": null, "person_id": 10, "modified_by": null, "id": 153}', None, None, None, None, 150L, 537L), (154L, 'UserRole', 53283L, 'created', '{"display_name": "user1@ggrc.com <-> ProgramOwner", "context_id": 537, "created_at": "2015-07-17T20:45:54", "updated_at": "2015-07-17T20:45:54", "role_id": 1, "modified_by_id": null, "person_id": 31, "modified_by": null, "id": 154}', None, None, None, None, 150L, 537L), (611L, 'UserRole', 53283L, 'created', '{"display_name": "pp@pp.com <-> WorkflowOwner in Prasanna\'s test workflow", "context_id": 982, "created_at": "2015-07-29T17:55:24", "updated_at": "2015-07-29T17:55:24", "role_id": 18, "modified_by_id": null, "person_id": 6, "modified_by": null, "id": 611}', None, None, None, None, 150L, 982L), (1005L, 'UserRole', 53283L, 'created', '{"display_name": "user10@example.com <-> Creator", "context_id": null, "created_at": "2015-08-13T14:28:09", "updated_at": "2015-08-13T14:28:09", "role_id": 17, "modified_by_id": null, "person_id": 70, "modified_by": null, "id": 1005}', None, None, None, None, 150L, None)  ... displaying 10 of 675 total bound parameter sets ...  (4567L, 'UserRole', 53283L, 'created', '{"display_name": "gg@gg.com <-> ProgramOwner in NIST", "context_id": 3241, "created_at": "2016-06-17T11:56:32", "updated_at": "2016-06-17T11:56:32", "role_id": 1, "modified_by_id": null, "person_id": 194, "modified_by": null, "id": 4567}', None, None, None, None, 150L, 3241L), (4587L, 'UserRole', 53283L, 'created', '{"display_name": "aa@aaa.com <-> Creator", "context_id": null, "created_at": "2016-08-17T22:44:15", "updated_at": "2016-08-17T22:44:15", "role_id": 17, "modified_by_id": null, "person_id": 235, "modified_by": null, "id": 4587}', None, None, None, None, 150L, None))
```

And I'm also really really afraid of this:

```
Total queries: 199614
```

alieh-rymasheuski commented:
The problem with UserRoles is resolved with the migration from #4824.

I'll check if I can reduce the number of DB requests.

alieh-rymasheuski commented:
The number of queries was O(number of updated revisions) even before my changes. Decreasing the number of queries will take too much time; let's tackle this in a separate PR (I created a separate ticket GGRC-609 for this optimization).

The problem with invalid dates was fixed with explicit calls to SQL `now()`.

Putting this PR `on hold` until #4824 is merged.

Smotko commented:
I was trying to run this on the latest grc-dev database, but it failed without an explicit error (I used launch_gae_ggrc):

```
INFO     2016-12-22 12:56:49,024 revisions.py:177] Updating revisions for: Relationship
INFO     2016-12-22 12:59:09,906 module.py:627] default: "POST /_background_tasks/refresh_revisions HTTP/1.1" 500 197298
WARNING  2016-12-22 12:59:09,943 taskqueue_stub.py:1983] Task refresh_revisions1482411214_1060 failed to execute. The task has no remaining retries. Failing permanently after 0 retries and 335 seconds
```

The python process was using a lot of RAM while the background task was running:
![screen shot 2016-12-22 at 13 57 33](https://cloud.githubusercontent.com/assets/513444/21426468/5d0cb2fc-c84f-11e6-96b5-3b6960f484f8.png)

I have 2GB available in my virtual box so the issue might be that it ran out. The max amount of RAM an appengine instance can have is 1024 MB so I am really worried about performance of this even if we ignore the amount of SQL statement that ran...

alieh-rymasheuski commented:
@Smotko I updated the code; now when we try loading fewer objects, the peak memory consumption on my local machine is about 1.0x Gib. I think it is still possible to reduce this even further.

As for many requests: on a local run, a single call to recreate revisions created 160697 revisions (141954 non-delete revisions, so each of them effectively took a separate `INSERT` statement) and updated 64481 revisions (each of them took a separate `UPDATE` statement). I don't see a valid way to reduce the number of requests.

Serialization is done with the data fetched by `eager_query` and emits a O(1) number of DB queries for each single model.

alieh-rymasheuski commented:
To reduce the memory consumption, the following can be done:
1. Create a temporary table with `resource_id, revision_id` pairs resulted from FULL OUTER JOIN for an object type.
2. For each `resource_id IS NULL, revision_id IS NOT NULL` create a `"deleted"` revision.
3. For each `resource_id IS NOT NULL, revision_id IS NOT NULL` perform a revision update.
4. For each `resource_id IS NOT NULL, revision_id IS NULL` create a `"update"`/`"create"` revision.

This four steps should be performed for each `type in valid_types`.
Step 2 can be performed in the DB same as now.
Steps 3 and 4 can be performed with JSON generation in Python code. We should fetch the object ids in chunks of, say, 1000 from the temporary table, fetch the objects themselves with `eager_query`, do the `INSERT`/`UPDATE`, repeat. This will reduce the memory consumption to a predictable value not really depending on the number of objects.

If the memory consumption in the current implementation (around 1 Gib) is OK, this optimization should be performed at a later stage.

egorhm commented:
@alieh-rymasheuski conflicts needs to be resolved.

alieh-rymasheuski commented:
@Smotko These warnings are caused by data inconsistencies in our DB. For instance, `ObjectPerson.Risk_personable` is not defined properly, or there are `ObjectOwner` instances for long gone objects, or `UserRole.person_id` is not a foreign key to `Person`. These fixes are definitely not in scope of this PR.

I reported several tickets: GGRC-747, GGRC-748 and GGRC-749 to handle it at some point.

alieh-rymasheuski commented on src/ggrc/utils/revisions.py:
```diff
@@ -3,91 +3,191 @@
+  valid_types = {model.__name__ for model in all_models.all_models}
+
+  # No sense in storing revisions for Revisions or Events
+  valid_types -= {"Revision", "Event"}
```
Done.

Smotko commented:
@alieh-rymasheuski those tickets need to be resolved before the prod release imo or can you guarantee that there will be no user facing bugs from these warnings that we see on grc-dev and warnings that we might get in prod? If you can guarantee this I see no problem merging, but the fact the number of warnings is even too big to process makes me very nervous.

alieh-rymasheuski commented:
@Smotko The code from this PR does not introduce new bugs related to these warnings.

The warnings are reported when generating JSONs for revisions for:
1. Mappings - not used anywhere
- ObjectPerson
- TaskGroupObject
- ObjectDocument
- ObjectOwner
- ObjectFolder

2. Request - deprecated model not shown on the frontend
3. UserRole - not used anywhere as well

Smotko commented:
What do you mean that user roles are not used anywhere? All our RBAC depends on them. If the revisions for user roles are wrong we might not be able to figure out how someone got access to a program/workflow and that could be really bad.

Smotko commented:
Merging for now, but I really hope we reduce the verbosity of this function before running in prod :)
